yarn/src/main/scala/org/apache/spark/scheduler/cluster/SchedulerExtensionService.scala

```diff
@@ -0,0 +1,137 @@
+   */
+  def start(binding: SchedulerExtensionServiceBinding): Unit
+
+  /**
+   * Stop the service
+   * The `stop()` operation MUST be idempotent, and succeed even if `start()` was
+   * never invoked.
+   */
+  def stop(): Unit
+}
+
+/**
+ * Binding information for a [[SchedulerExtensionService]]
+ * @param sparkContext current spark context
+ * @param applicationId YARN application ID
+ * @param attemptId optional AttemptID.
+ */
+case class SchedulerExtensionServiceBinding(
+    sparkContext: SparkContext,
+    applicationId: ApplicationId,
+    attemptId: Option[ApplicationAttemptId] = None)
+
+/**
+ * Container for [[SchedulerExtensionService]] instances.
+ *
+ * Loads Extension Services from the configuration property
+ * `"spark.yarn.services"`, instantiates and starts them.
+ * When stopped, it stops all child entries.
+ *
+ * The order in which child extension services are started and stopped
+ * is undefined.
+ *
+ */
+private[spark] class SchedulerExtensionServices extends SchedulerExtensionService
+    with Logging {
+  private var services: List[SchedulerExtensionService] = Nil
+  private var sparkContext: SparkContext = _
+  private var appId: ApplicationId = _
+  private var attemptId: Option[ApplicationAttemptId] = _
+  private val started = new AtomicBoolean(false)
+  private var binding: SchedulerExtensionServiceBinding = _
+
+  /**
+   * Binding operation will load the named services and call bind on them too; the
+   * entire set of services are then ready for `init()` and `start()` calls
+
+   * @param binding binding to the spark application and YARN
+   */
+  def start(binding: SchedulerExtensionServiceBinding): Unit = {
+    if (started.getAndSet(true)) {
+      logWarning("Ignoring re-entrant start operation")
+      return
+    }
+    require(binding.sparkContext != null, "Null context parameter")
+    require(binding.applicationId != null, "Null appId parameter")
+    this.binding = binding
+    sparkContext = binding.sparkContext
+    appId = binding.applicationId
+    attemptId = binding.attemptId
+    logInfo(s"Starting Yarn extension services with app ${binding.applicationId}" +
+        s" and attemptId $attemptId")
+
+    services = sparkContext.getConf.getOption(SchedulerExtensionServices.SPARK_YARN_SERVICES)
+        .map { s =>
+      s.split(",").map(_.trim()).filter(!_.isEmpty)
+        .map { sClass =>
+            val instance = Utils.classForName(sClass)
+                .newInstance()
+                .asInstanceOf[SchedulerExtensionService]
+            // bind this service
+            instance.start(binding)
+            logInfo(s"Service $sClass started")
+            instance
+          }
+    }.map(_.toList).getOrElse(Nil)
+  }
+
+  /**
+   * Get the list of services
+   * @return a list of services; Nil until the service is started
+   */
+  def getServices: List[SchedulerExtensionService] = {
+    services
+  }
+
+  override def stop(): Unit = {
+    logInfo(s"Stopping $this")
```

Comment:
good catch. will set & skip the operation if !started. This will permit reissuing of the start() call from a stopped service, but a full state machine here is probably overkill.